src/units/attack_type.cpp

```diff
@@ -96,12 +117,51 @@ std::string attack_type::accuracy_parry_description() const
 	return s.str();
 }
 
+namespace {
+/**
+ * Print "Recursion limit reached" log messages, including deduplication if the same problem has
+ * already been logged.
+ */
+void show_recursion_warning(const attack_type& attack, const config& filter) {
+	// This function is only called when the recursion limit has already been reached, meaning the
+	// filter has already been parsed multiple times, so I'm not trying to optimize the performance
+	// of this; it's merely to prevent the logs getting spammed. For example, each of
+	// four_cycle_recursion_branching and event_test_filter_attack_student_weapon_condition only log
+	// 3 unique messages, but without deduplication they'd log 1280 and 392 respectively.
+	static std::vector<std::tuple<std::string, std::string>> already_shown;
+
+	auto identifier = std::tuple<std::string, std::string>{attack.id(), filter.debug()};
+	if(utils::contains(already_shown, identifier)) {
+		return;
+	}
+
+	std::string_view filter_text_view = std::get<1>(identifier);
+	utils::trim(filter_text_view);
+	ERR_UT << "Recursion limit reached for weapon '" << attack.id()
+		<< "' while checking filter '" << filter_text_view << "'";
+
+	// Arbitrary limit, just ensuring that having a huge number of specials causing recursion
+	// warnings can't lead to unbounded memory consumption here.
+	if(already_shown.size() > 100) {
+		already_shown.clear();
+	}
+	already_shown.push_back(std::move(identifier));
+}
+
 /**
  * Returns whether or not *this matches the given @a filter, ignoring the
  * complexities introduced by [and], [or], and [not].
  */
-static bool matches_simple_filter(const attack_type & attack, const config & filter, const std::string& tag_name)
+bool matches_simple_filter(const attack_type& attack, const config& filter, const std::string& tag_name)
 {
+	//update and check variable_recursion for prevent check special_id/type_active in case of infinite recursion.
+	attack_type::recursion_guard filter_lock;
+	filter_lock = attack.update_variables_recursion();
```

Comment:
I don't understand why this is two lines instead of one? Is there a need to default-initialize and then assign? It doesn't look like default-initializing has any useful logic in it.

Reply:
I'll fix it, that's a leftover from a version where the assignment was within a conditional block.